go/vt/vtgate/engine/memory_sort.go

```diff
@@ -0,0 +1,215 @@
+		result.RowsAffected = uint64(count)
+	}
+	return result, nil
+}
+
+// StreamExecute satisfies the Primtive interface.
+func (ms *MemorySort) StreamExecute(vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool, callback func(*sqltypes.Result) error) error {
+	count, err := ms.fetchCount(bindVars)
+	if err != nil {
+		return err
+	}
+
+	// You have to reverse the ordering because the highest values
+	// must be dropped once the upper limit is reached.
+	sh := &sortHeap{
+		orderBy: ms.OrderBy,
+		reverse: true,
+	}
+	err = ms.Input.StreamExecute(vcursor, bindVars, wantfields, func(qr *sqltypes.Result) error {
+		if len(qr.Fields) != 0 {
+			if err := callback(&sqltypes.Result{Fields: qr.Fields}); err != nil {
+				return err
+			}
+		}
+		for _, row := range qr.Rows {
+			heap.Push(sh, row)
+		}
+		for len(sh.rows) > count {
+			_ = heap.Pop(sh)
+		}
+		if len(sh.rows) > vcursor.MaxMemoryRows() {
+			return fmt.Errorf("in-memory row count exceeded allowed limit of %d", vcursor.MaxMemoryRows())
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	if sh.err != nil {
+		return sh.err
+	}
+	// Set ordering to normal for the final ordering.
+	sh.reverse = false
+	sort.Sort(sh)
+	if sh.err != nil {
+		// Unreachable.
+		return sh.err
+	}
+	return callback(&sqltypes.Result{Rows: sh.rows})
+}
+
+// GetFields satisfies the Primtive interface.
+func (ms *MemorySort) GetFields(vcursor VCursor, bindVars map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
+	return ms.Input.GetFields(vcursor, bindVars)
+}
+
+func (ms *MemorySort) fetchCount(bindVars map[string]*querypb.BindVariable) (int, error) {
+	resolved, err := ms.UpperLimit.ResolveValue(bindVars)
+	if err != nil {
+		return 0, err
+	}
+	if resolved.IsNull() {
+		return math.MaxInt64, nil
+	}
+	num, err := sqltypes.ToUint64(resolved)
+	if err != nil {
+		return 0, err
+	}
+	count := int(num)
+	if count < 0 {
+		return 0, fmt.Errorf("requested limit is out of range: %v", num)
+	}
+	return count, nil
+}
+
+// sortHeap is sorted based on the orderBy params.
+// Implementation is similar to scatterHeap
+type sortHeap struct {
+	rows    [][]sqltypes.Value
+	orderBy []OrderbyParams
+	reverse bool
+	err     error
+}
+
+func (sh *sortHeap) Len() int {
+	return len(sh.rows)
+}
+
+func (sh *sortHeap) Less(i, j int) bool {
+	for _, order := range sh.orderBy {
+		if sh.err != nil {
+			return true
+		}
+		cmp, err := sqltypes.NullsafeCompare(sh.rows[i][order.Col], sh.rows[j][order.Col])
+		if err != nil {
+			sh.err = err
+			return true
+		}
+		if cmp == 0 {
+			continue
+		}
+		// This is equivalent to:
+		//if !sj.reverse {
+		//	if order.Desc {
+		//		cmp = -cmp
+		//	}
+		//} else {
+		//	if !order.Desc {
+		//		cmp = -cmp
+		//	}
+		//}
+		if sh.reverse != order.Desc {
+			cmp = -cmp
+		}
+		return cmp < 0
+	}
+	return true
+}
+
+func (sh *sortHeap) Swap(i, j int) {
+	sh.rows[i], sh.rows[j] = sh.rows[j], sh.rows[i]
+}
+
+func (sh *sortHeap) Push(x interface{}) {
```

Comment:
From the tests, it looks fine, but I don't understand how this heap works. I don't see how you make the last element of the array be the max (or min) with a simple append.

Reply:
Yeah. Heap is cool. It lazily rearranges items O(log N), and ensures that the highest value is always at the top.

Reply:
Oh, you just had to implement the right interface :)

Reply:
can you annotate the interface methods with a comment specifying which interface they satisfy?